src/main/java/org/opensearch/flowframework/workflow/WorkflowProcessSorter.java

```diff
@@ -74,17 +75,68 @@ public List<ProcessNode> sortProcessNodes(Workflow workflow) {
                 .collect(Collectors.toList());
 
             TimeValue nodeTimeout = parseTimeout(node);
-            ProcessNode processNode = new ProcessNode(node.id(), step, data, predecessorNodes, threadPool, nodeTimeout);
+            ProcessNode processNode = new ProcessNode(
+                node.id(),
+                step,
+                node.previousNodeInputs(),
+                data,
+                predecessorNodes,
+                threadPool,
+                nodeTimeout
+            );
             idToNodeMap.put(processNode.id(), processNode);
             nodes.add(processNode);
         }
 
         return nodes;
     }
 
+    /**
+     * Validates a sorted workflow, determines if each process node's user inputs and predecessor outputs match the expected workflow step inputs
+     * @param processNodes A list of process nodes
+     * @throws Exception on validation failure
+     */
+    public void validateGraph(List<ProcessNode> processNodes) throws Exception {
+
+        WorkflowValidator validator = WorkflowValidator.parse("mappings/workflow-steps.json");
+
+        // Iterate through process nodes in graph
+        for (ProcessNode processNode : processNodes) {
+
+            // Get predecessor nodes types of this processNode
+            List<String> predecessorNodeTypes = processNode.predecessors()
+                .stream()
+                .map(x -> x.workflowStep().getName())
+                .collect(Collectors.toList());
+
+            // Compile a list of outputs from the predecessor nodes based on type
+            List<String> predecessorOutputs = new ArrayList<>();
+            for (String nodeType : predecessorNodeTypes) {
+                List<String> nodeTypeOutputs = validator.getWorkflowStepValidators().get(nodeType).getOutputs();
+                predecessorOutputs.addAll(nodeTypeOutputs);
+            }
+
+            // Retrieve all the user input data from this node
+            List<String> currentNodeUserInputs = new ArrayList<String>(processNode.input().getContent().keySet());
+
+            // Combine both predecessor outputs and current node user inputs
+            List<String> allInputs = new ArrayList<>();
+            allInputs.addAll(predecessorOutputs);
+            allInputs.addAll(currentNodeUserInputs);
+
```

Comment:
123-125 can be written as:

```java
List<String> allInputs =  Stream.concat(predecessorOutputs.stream(), currentNodeUserInputs.stream())
                                        .collect(Collectors.toList());

Reply:
This looks cleaner, thanks for the suggestion